### ISSUE TITLE
How to find the number of epochs a model was trained for?

### ISSUE THREAD

**Issue by angadgill** [fchollet/keras#1766]
Given a trained Keras model, is there a way to check how many epochs were used to train it? For example, `print model.nb_epoch`. 
When training a model, say, 20 epochs at a time this will help figure out how many total epochs it has been trained on.

**Comment by KeironO** [comment]
I guess you're attempting to perform some sort of gridsearch to find the optimum training parameters to use?

As far as I am aware, there is nothing about the fitting of the NN stored in the saved models. You'd have to do this manually.

**Comment by angadgill** [comment]
Okay, thanks. I guess I'll have to track it manually. 

Any thoughts on adding `model.nb_epoch` to the class?

**Comment by fchollet** [comment]
> Any thoughts on adding model.nb_epoch to the class?

The number of "epochs" a model was trained for only makes sense in a few situations. It assume you use "epochs" (what if you train with randomly generated batches out of a data generator?). What if you use several datasets? What if you use pre-training? etc.

So: no, we won't do that. A model is just a state.

**Comment by angadgill** [comment]
That makes sense. Thanks for the response @fchollet!

**Comment by CyrilWendl** [comment]
I could still see situations where it would be helpful to see stats on the number of epochs from a model:
1. In case you decide to stop training, save your temporary model and continue later, continue the epoch counter at the total number of epochs during which the model was previously trained.
2. For equivalent model setups where callbacks like ReduceLnRate or EarlyStopping were used, the number of epochs could be an indicator of how early the model stopped training.

**Comment by rakoczUCLA** [comment]
you can save the length of your model's history when training it.

`r = model.fit(X_train,y_train,epochs=5)
n_epochs = len(r.history['loss'])`

**Comment by hoefling** [comment]
There is a `CSVLogger` to persist the history:

```py
logdir = pathlib.Path("logs", ...)
model.fit(
        ...,
        callbacks=[tf.keras.callbacks.CSVLogger(logdir / "history.csv", append=True)],
)
```
Example contents of `history.csv` afterwards:
```csv
epoch,learning rate,loss,rmse,val_loss,val_rmse
0,1e-06,2.688473062008105e-18,1.6396568547349943e-09,3.596072587303844e-18,1.896330203976504e-09
1,1e-06,3.3348301315006625e-18,1.826152340456133e-09,3.596072587303844e-18,1.896330203976504e-09
2,1e-06,3.282354413825758e-18,1.8117276567863883e-09,3.596072587303844e-18,1.896330203976504e-09
3,1e-06,3.199610709971466e-18,1.7887455960874377e-09,3.576532099693504e-18,1.891170997581071e-09
```
Reading it:
```py
import csv
import pathlib
from pprint import pprint

with pathlib.Path("history.csv").open() as fp:
     data = list(csv.DictReader(fp)) 

pprint(data)
```
Output:
```py
[{'epoch': '0',
  'learning rate': '1e-06',
  'loss': '2.688473062008105e-18',
  'rmse': '1.6396568547349943e-09',
  'val_loss': '3.596072587303844e-18',
  'val_rmse': '1.896330203976504e-09'},
 {'epoch': '1',
  'learning rate': '1e-06',
  'loss': '3.3348301315006625e-18',
  'rmse': '1.826152340456133e-09',
  'val_loss': '3.596072587303844e-18',
  'val_rmse': '1.896330203976504e-09'},
 {'epoch': '2',
  'learning rate': '1e-06',
  'loss': '3.282354413825758e-18',
  'rmse': '1.8117276567863883e-09',
  'val_loss': '3.596072587303844e-18',
  'val_rmse': '1.896330203976504e-09'},
 {'epoch': '3',
  'learning rate': '1e-06',
  'loss': '3.199610709971466e-18',
  'rmse': '1.7887455960874377e-09',
  'val_loss': '3.576532099693504e-18',
  'val_rmse': '1.891170997581071e-09'}]
```

**Comment by NoraAl** [comment]
Epochs are there only to put an **upper bound** on the number of times you passed through the whole dataset. That is it! Well, you can also use an infinite loop and put a condition, say certain accuracy, to break the training without explicitly specifying the number of epochs. 

However, they can show how many **passes it took a model to move weights from their initial state to another state**. That can be helpful in judging two initialization schemes. 
Remember, a model trained for 1000 epochs can be exactly another model trained only for 2 epochs. You can not tell the difference. 

**For example**, say you have two identical models with a single weight, one with weight of `2`, and another with weight of `99`. Let's also say that the perfect weight should be `1`. You can see that the first model is most likely to take fewer epochs to reach the goal of `1` than the latter. Two models reach the same state after different numbers of epochs, which does not tell anything about their current state but rather their initial state.

**Comment by mauriciocramos** [comment]
> > Any thoughts on adding model.nb_epoch to the class?
> 
> The number of "epochs" a model was trained for only makes sense in a few situations. It assume you use "epochs", but what if you train with randomly generated batches out of a data generator? What if you use several datasets? What if you use pre-training? etc.
> 
> So: no, we won't do that. A model is just a state.

I'm facing one of those "few situations":
* I was given a simple pre-trained LSTM model, its architecture, optimizer parameters and training data from an author I don't have access to. 
* I would like to fit that model myself for *reproducibility reasons* as best as possible but the fit parameters such as `epochs` and `batch_size` are unknown.

**Comment by mauriciocramos** [comment]
> model.optimizer.iterations.numpy() // (len(x_train) // batch_size)

I believe that is a good initial approach for my case explained in my previous message:

> I have a pre-trained model, its architecture, optimizer parameters and training data but don't have the fit parameters such as `epochs`,  `batch_size` and any `EarlyStopping` callback.

Based on your approach and assuming the original fitting used the default `batch_size=32` and no `EarlyStopping`, I could infer the number of trained epochs as:

```
np.ceil(model.optimizer.iterations.numpy() / np.ceil(len(X_train) / batch_size))
```
